<h1><b>Demographic Data Analyzer</b></h1>
<p>By Shabri Chairul Akmal</p>
<hr/>
<h2><b>Project Overview</b></h2>
<p style="text-align: justify;">This notebook implements a demographic data analysis function that processes the UCI Adult Census Income dataset. The function performs various statistical calculations to understand demographic patterns and income distributions across different population groups.</p>
<hr/>
<h2><b>Problem Statement</b></h2>
<p style="text-align: justify;">In this project, we will act as a data analyst tasked with exploring a demographic dataset to uncover patterns related to income, education, and other socio-economic factors. Our goal is to use Python (specifically the Pandas and NumPy libraries) to perform data analysis and answer a series of specific questions:</p>
<ol>
<li><b>Racial Composition</b>
<p>How many of each race are represented in this dataset?</p></li>
<li><b>Age Demographics</b><p>What is the average age of men?</p></li>
<li><b>Educational Attainment</b><p>What is the percentage of people who have a Bachelor's degree?</p></li>
<li><b>Education vs. Income</b>
<ul>
<li>What percentage of people with advanced education (<code>Bachelors</code>, <code>Masters</code>, or <code>Doctorate</code>) make more than 50K?</li>
<li>What percentage of people without advanced education make more than 50K?</li>
</ul>
</li>
<li><b>Work Hours</b><p>What is the minimum number of hours a person works per week (hours-per-week feature)?</p></li>
<li><b>Work Hours vs. Income</b><p>What percentage of the people who work the minimum number of hours per week have a salary of >50K?</p></li>
<li><b>Origin Perspective</b><p>What origin country has the highest percentage of people that earn >50K?</p></li>
<li><b>Occupation insight</b><p>Identify the most popular occupation for those who earn >50K in India's origin.</p></li></ol>
<hr/>
<h2><b>Solution & Strategy</b></h2>
<p>The general strategy to fulfill the question will be:</p>
<ul>
<li>Load and explore data using Pandas</li>
<li>Filter and group data based on specific conditions</li>
<li>Perform percentage calculations using NumPy</li>
</ul>
<p>While the specific strategy for answering each question ishere below:</p>
<ol>
<li><b>Racial Composition</b>
<p><b>Strategy:</b> Use the <code>value_counts()</code> method on the 'race' column to count the occurrences of each race. This will return a Pandas Series with race names as index labels and counts as values.</p>
<p><b>Implementation:</b> <code>race_count = pd.Series(df['race'].value_counts())</code></p>
</li>

<li><b>Average Age of Men</b>
<p><b>Strategy:</b> Filter the dataset to include only rows where sex is 'Male', then calculate the mean of the 'age' column. Use NumPy's <code>round()</code> function to round to 1 decimal place.</p>
<p><b>Implementation:</b> <code>average_age_men = np.round(df[df['sex']=='Male']['age'].mean(),1)</code></p>
</li>

<li><b>Percentage with Bachelor's Degree</b>
<p><b>Strategy:</b> Count rows where education is 'Bachelors', divide by total rows, multiply by 100, and round to 1 decimal place.</p>
<p><b>Implementation:</b> <code>percentage_bachelors = np.round(len(df[df['education']=='Bachelors'])/len(df['education'])*100,1)</code></p>
</li>

<li><b>Education vs. Income</b>
<p><b>Strategy:</b> Create boolean masks for higher education (Bachelors, Masters, Doctorate) and lower education (all others). Then calculate the percentage of each group earning >50K by filtering with both conditions.</p>
<p><b>Implementation:</b></p>
<pre>
higher_education = df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])
lower_education = ~(df['education'].isin(['Bachelors', 'Masters', 'Doctorate']))

higher_education_rich = np.round(len(df[higher_education & (df['salary'] == '>50K')])/len(df[higher_education])*100,1)
lower_education_rich = np.round(len(df[lower_education & (df['salary'] == '>50K')])/len(df[lower_education])*100,1)
</pre>
</li>

<li><b>Minimum Work Hours</b>
<p><b>Strategy:</b> Use the <code>min()</code> function on the 'hours-per-week' column to find the minimum value.</p>
<p><b>Implementation:</b> <code>min_work_hours = df['hours-per-week'].min()</code></p>
</li>

<li><b>Work Hours vs. Income</b>
<p><b>Strategy:</b> First filter to get rows with minimum work hours. Then count how many of those have salary >50K. Calculate the percentage.</p>
<p><b>Implementation:</b></p>
<pre>
num_min_workers = df[(df['hours-per-week'] == min_work_hours) & (df['salary'] == '>50K')]
rich_percentage = np.round(len(num_min_workers)/len(df[df['hours-per-week'] == min_work_hours])*100,1)
</pre>
</li>

<li><b>Origin country with Highest Percentage of >50K Earners</b>
<p><b>Strategy:</b> For each origin country, divide the count of >50K earners by the total count of people from that country. Sort descending and take the top country.</p>
<p><b>Implementation:</b></p>
<pre>
highest_earning = (np.round(df[df['salary'] == '>50K']['native-country'].value_counts()/df['native-country'].value_counts()*100,1)).sort_values(ascending = False)
highest_earning_country = highest_earning.index[0]
highest_earning_country_percentage = highest_earning.iloc[0]
</pre>
</li>

<li><b>Most Popular Occupation with India's Origin for >50K Earners</b>
<p><b>Strategy:</b> Filter the data for India and salary >50K, then use <code>value_counts()</code> on the 'occupation' column and take the first index.</p>
<p><b>Implementation:</b> <code>top_IN_occupation = (df[(df['salary'] == '>50K') & (df['native-country'] == 'India')]['occupation'].value_counts()).index[0]</code></p>
</li>
</ol>
<hr/>
<h2><b>Implementation</b></h2>
<h3>Cell 1: Import Libraries</h3>




In [1]:
import pandas as pd
import numpy as np

<h3>Cell 2: Import CSV file</h3>

In [2]:
url = "https://shabriakmal.github.io/Data-Analysis-python-freeCodeCamp/Demographic_Data_Analyzer/adult.data.csv"
df = pd.read_csv(url)

<h3>Cell 3: The Complete Solution Function</h3>

In [3]:
def calculate_demographic_data(print_data=True):

    # How many of each race are represented in this dataset? This should be a Pandas series with race names as the index labels.
    race_count = pd.Series(df['race'].value_counts())

    # What is the average age of men?
    average_age_men = np.round(df[df['sex']=='Male']['age'].mean(),1)

    # What is the percentage of people who have a Bachelor's degree?
    percentage_bachelors = np.round(len(df[df['education']=='Bachelors'])/len(df['education'])*100,1)

    # What percentage of people with advanced education (`Bachelors`, `Masters`, or `Doctorate`) make more than 50K?
    # What percentage of people without advanced education make more than 50K?

    # with and without `Bachelors`, `Masters`, or `Doctorate`
    higher_education = df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])
    lower_education = ~(df['education'].isin(['Bachelors', 'Masters', 'Doctorate']))

    # percentage with salary >50K
    higher_education_rich = np.round(len(df[higher_education & (df['salary'] == '>50K')])/len(df[higher_education])*100,1)
    lower_education_rich = np.round(len(df[lower_education & (df['salary'] == '>50K')])/len(df[lower_education])*100,1)

    # What is the minimum number of hours a person works per week (hours-per-week feature)?
    min_work_hours = df['hours-per-week'].min()

    # What percentage of the people who work the minimum number of hours per week have a salary of >50K?
    num_min_workers = df[(df['hours-per-week'] == min_work_hours) & (df['salary'] == '>50K')]

    rich_percentage = np.round(len(num_min_workers)/len(df[df['hours-per-week'] == min_work_hours])*100,1)

    # What country has the highest percentage of people that earn >50K?
    highest_earning = (np.round(df[df['salary'] == '>50K']['native-country'].value_counts()/df['native-country'].value_counts()*100,1)).sort_values(ascending = False)
    highest_earning_country = highest_earning.index[0]
    highest_earning_country_percentage = highest_earning.iloc[0]

    # Identify the most popular occupation for those who earn >50K in India.
    top_IN_occupation = (df[(df['salary'] == '>50K') & (df['native-country'] == 'India')]['occupation'].value_counts()).index[0]

    # DO NOT MODIFY BELOW THIS LINE

    if print_data:
        print("Number of each race:\n", race_count)
        print("Average age of men:", average_age_men)
        print(f"Percentage with Bachelors degrees: {percentage_bachelors}%")
        print(f"Percentage with higher education that earn >50K: {higher_education_rich}%")
        print(f"Percentage without higher education that earn >50K: {lower_education_rich}%")
        print(f"Min work time: {min_work_hours} hours/week")
        print(f"Percentage of rich among those who work fewest hours: {rich_percentage}%")
        print("Country with highest percentage of rich:", highest_earning_country)
        print(f"Highest percentage of rich people in country: {highest_earning_country_percentage}%")
        print("Top occupations in India:", top_IN_occupation)

    return {
        'race_count': race_count,
        'average_age_men': average_age_men,
        'percentage_bachelors': percentage_bachelors,
        'higher_education_rich': higher_education_rich,
        'lower_education_rich': lower_education_rich,
        'min_work_hours': min_work_hours,
        'rich_percentage': rich_percentage,
        'highest_earning_country': highest_earning_country,
        'highest_earning_country_percentage':
        highest_earning_country_percentage,
        'top_IN_occupation': top_IN_occupation
    }

<h3>Cell 4: Get the Result</h3>

In [4]:
calculate_demographic_data(print_data=True)

Number of each race:
 race
White                 27816
Black                  3124
Asian-Pac-Islander     1039
Amer-Indian-Eskimo      311
Other                   271
Name: count, dtype: int64
Average age of men: 39.4
Percentage with Bachelors degrees: 16.4%
Percentage with higher education that earn >50K: 46.5%
Percentage without higher education that earn >50K: 17.4%
Min work time: 1 hours/week
Percentage of rich among those who work fewest hours: 10.0%
Country with highest percentage of rich: Iran
Highest percentage of rich people in country: 41.9%
Top occupations in India: Prof-specialty


{'race_count': race
 White                 27816
 Black                  3124
 Asian-Pac-Islander     1039
 Amer-Indian-Eskimo      311
 Other                   271
 Name: count, dtype: int64,
 'average_age_men': np.float64(39.4),
 'percentage_bachelors': np.float64(16.4),
 'higher_education_rich': np.float64(46.5),
 'lower_education_rich': np.float64(17.4),
 'min_work_hours': 1,
 'rich_percentage': np.float64(10.0),
 'highest_earning_country': 'Iran',
 'highest_earning_country_percentage': np.float64(41.9),
 'top_IN_occupation': 'Prof-specialty'}

<h3>Cell 5: Invalid Input Test</h3>

In [5]:
calculate_demographic_data(print_data=False)

{'race_count': race
 White                 27816
 Black                  3124
 Asian-Pac-Islander     1039
 Amer-Indian-Eskimo      311
 Other                   271
 Name: count, dtype: int64,
 'average_age_men': np.float64(39.4),
 'percentage_bachelors': np.float64(16.4),
 'higher_education_rich': np.float64(46.5),
 'lower_education_rich': np.float64(17.4),
 'min_work_hours': 1,
 'rich_percentage': np.float64(10.0),
 'highest_earning_country': 'Iran',
 'highest_earning_country_percentage': np.float64(41.9),
 'top_IN_occupation': 'Prof-specialty'}

<hr/>
<h2><b>Output Interpretation</b></h2>
<p>From the result obtained,the conclusion will be:</p>
<ol>
<li><b>Racial Composition</b></li>
<p>The working race dominated by <b>white</b> race, followed by <b>black</b>, <b>asian-pac-islander</b>, <b>amer-indian-eskimo</b>, and the others. It can represent <b>racial composition</b> as a whole, not just working group or it can represent the <b>opportunity</b> given to each race.</p>
<li><b>Age Demographics</b></li>
<p>Average age of working men: <b>39.4 years</b>. The typical male in the dataset is in their late 30s, likely in their prime earning years.</p>
<li><b>Educational Attainment</b></li>
<p>Only 16.4% of the population holds a Bachelor's degree. This indicates a relatively <b>low level</b> of higher education attainment overall. Most people in the dataset do not have a college degree, which could be a barrier to higher-paying jobs.</p>
<li><b>Education vs. Income</b></li>
<table border="1">
  <thead style="text-align: left;">
    <tr>
      <th>Education Level</th>
      <th>% Earning >50K</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Advanced (Bachelors/Masters/Doctorate)</td>
      <td>46.5%</td>
    </tr>
    <tr>
      <td>Without Advanced Education</td>
      <td>17.4%</td>
    </tr>
  </tbody>
</table>
<p> From the data the percentage of high degree holder earn >50K is <b>46.5%</b>  vs. <b>17.4%</b> without. People with advanced education are <b>2.7x</b> more likely to earn >50K. This means education is a <b>strong predictor of high income</b> </p>
<li><b> Work Hours Analysis</b></li>
<p>Based on result, minimum work hours forworking men is <b>1 hour/week</b>. And only <b>10%</b> of those working minimum hours earn >50K. While most people work 40 hours/week (standard full-time). This means people working very few hours are rarely high earners.</p>
<li><b>Country of Origin Analysis</b></li>
<p>People with <b>Iran</b> Nationality as origin country has the highest percentage of high earners with <b>41.9%</b> of its working people. This outperforms many other countries, including the USA. This means country of origin significantly affects earning potential. It also means that Iran has better education system since its high earner also has higher education.</p>
<li><b>Occupational Insight with India's Origin Case</b></li>
<p>Most popular occupation for high earners (>50K) from India: <b>Prof-specialty</b> (Professional Specialty). This suggests that for India's people, <b>professional occupations</b> are the path to high income. This likely includes roles like doctors, engineers, lawyers, and scientists.</p>
</ol>

In [4]:
!jupyter nbconvert Demographic_Data_Analyzer.ipynb --to html --lab full

[NbConvertApp] WARNING | Unrecognized alias: 'lab', it will have no effect.
[NbConvertApp] Converting notebook Demographic_Data_Analyzer.ipynb to html
[NbConvertApp] Writing 308162 bytes to Demographic_Data_Analyzer.html
